In [2]:
pip install ucimlrepo

Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd 
import numpy as np
from ucimlrepo import fetch_ucirepo

In [4]:
# fetch dataset 
mushroom = fetch_ucirepo(id=73)

# data (as pandas dataframes) 
X = mushroom.data.features 
y = mushroom.data.targets 
  
# metadata 
#print(mushroom.metadata) 
  
# variable information 
print(mushroom.variables) 

df=mushroom.data.original
df.head()

                        name     role         type demographic  \
0                  poisonous   Target  Categorical        None   
1                  cap-shape  Feature  Categorical        None   
2                cap-surface  Feature  Categorical        None   
3                  cap-color  Feature       Binary        None   
4                    bruises  Feature  Categorical        None   
5                       odor  Feature  Categorical        None   
6            gill-attachment  Feature  Categorical        None   
7               gill-spacing  Feature  Categorical        None   
8                  gill-size  Feature  Categorical        None   
9                 gill-color  Feature  Categorical        None   
10               stalk-shape  Feature  Categorical        None   
11                stalk-root  Feature  Categorical        None   
12  stalk-surface-above-ring  Feature  Categorical        None   
13  stalk-surface-below-ring  Feature  Categorical        None   
14    stal

cap-shape cap-surface cap-color bruises odor gill-attachment gill-spacing  \
0         x           s         n       t    p               f            c   
1         x           s         y       t    a               f            c   
2         b           s         w       t    l               f            c   
3         x           y         w       t    p               f            c   
4         x           s         g       f    n               f            w   

  gill-size gill-color stalk-shape  ... stalk-color-above-ring  \
0         n          k           e  ...                      w   
1         b          k           e  ...                      w   
2         b          n           e  ...                      w   
3         n          n           e  ...                      w   
4         b          k           t  ...                      w   

  stalk-color-below-ring veil-type veil-color ring-number ring-type  \
0                      w         p          w           o         p   
1                      w         p          w           o         p   
2                      w         p          w           o         p   
3                      w         p          w           o         p   
4                      w         p          w           o         e   

  spore-print-color population habitat poisonous  
0                 k          s       u         p  
1                 n          n       g         e  
2                 n          n       m         e  
3                 k          s       u         p  
4                 n          a       g         e  

[5 rows x 23 columns]

In [5]:
df.columns

Index(['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat', 'poisonous'],
      dtype='object')

In [6]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
label_encoder.fit(df['poisonous'])
encoded_data = label_encoder.transform(df['poisonous'])

#encodes each column in the dataset
for column in df.columns:
    df[column] = label_encoder.fit_transform(df[column])


In [7]:
test = [df.columns]
print(test)

[Index(['cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat', 'poisonous'],
      dtype='object')]


In [8]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [9]:
temp = df.drop('poisonous', axis=1, inplace=False)

In [10]:
X = df[df.drop('poisonous', axis=1, inplace=False).columns]
y = df['poisonous']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

# train a logistic regression model on the training set
from sklearn.naive_bayes import GaussianNB

# instantiate model
nb = GaussianNB()

# fit model
nb.fit(X_train, y_train)




GaussianNB()

In [11]:
y_pred_class = nb.predict(X_test)
y_pred_class

array([0, 0, 0, ..., 1, 0, 0])

**Classification accuracy**: percentage of correct predictions

In [12]:
# calculate accuracy
print(metrics.accuracy_score(y_test, y_pred_class))

0.916923076923077


In [13]:
df2 = pd.concat([X_test, y_test], axis=1)
df2 = pd.concat([df2.reset_index(), pd.Series(y_pred_class)], axis=1)
df2.head()

index  cap-shape  cap-surface  cap-color  bruises  odor  gill-attachment  \
0   1392          2            2          8        0     5                1   
1   4051          5            0          5        0     1                1   
2   3725          5            2          8        0     1                1   
3   7177          3            2          4        0     2                1   
4    103          5            3          9        1     0                1   

   gill-spacing  gill-size  gill-color  ...  stalk-color-below-ring  \
0             1          0           3  ...                       7   
1             1          1           9  ...                       7   
2             1          1           2  ...                       7   
3             0          1           0  ...                       7   
4             0          0           5  ...                       7   

   veil-type  veil-color  ring-number  ring-type  spore-print-color  \
0          0           2            1          0                  2   
1          0           2            1          4                  3   
2          0           2            1          4                  3   
3          0           2            1          0                  7   
4          0           2            1          4                  2   

   population  habitat  poisonous  0  
0           0        1          0  0  
1           4        0          1  0  
2           4        0          1  0  
3           4        2          1  1  
4           5        4          0  1  

[5 rows x 25 columns]

**Confusion matrix**: a table that describes the performance of a classification model

In [14]:
#first argument true values, second argument predicted values
print(metrics.confusion_matrix(y_test, y_pred_class))

[[759  61]
 [ 74 731]]


|            |Predicted: NO|Predicted: YES|
|------------|-------------|-------------|
|Actual: NO  |759 (TN)     |61 (FP)      |
|Actual: YES |74 (FN)      |731 (TP)     |


* True Negatives (TN): correctly predicted that the mushroom isnt poisonous: 759
* False Negatives (FN): incorrectly predicted that the mushroom isnt poisonous: 74
* False Positives (FP): incorrectly predicted that the mushroom is poisonous: 61
* True Positives (TP): correctly predicted that the mushroom is poisonous: 731

In [15]:
# slice confusion matrix into four pieces
confusion = metrics.confusion_matrix(y_test, y_pred_class)
print(confusion)
#[row, column]
TP = confusion[1, 1]
TN = confusion[0, 0]
FP = confusion[0, 1]
FN = confusion[1, 0]

[[759  61]
 [ 74 731]]


In [16]:
# Classification Accuracy
# overal, how often does it CORRECTLY predict that mushrooms are poisonous/not poisonous?

# use float to perform true division, not integer division (that rounds off)
# 2 ways to do 
print((TP + TN) / float(TP + TN + FP + FN))
print(metrics.accuracy_score(y_test, y_pred_class))

0.916923076923077
0.916923076923077


In [17]:
# Classification Error
# overall, how often does it INCORRECTLY predict that mushrooms are poisonous/not poisonous?

print((FP + FN) / float(TP + TN + FP + FN))
print(1 - metrics.accuracy_score(y_test, y_pred_class))

0.08307692307692308
0.08307692307692305


**Sensitivity**: When the actual value is positive, how often is the prediction correct?

### how often is it **CORRECTLY** predicted that the mushroom is **POISONOUS**?

* maximize this value
* aka "Recall" and "True Positive Rate"
* TP / all positive
    * all positive = TP + FN

In [18]:
sensitivity = TP / float(FN + TP)

print(sensitivity)
print(metrics.recall_score(y_test, y_pred_class))

0.9080745341614906
0.9080745341614906


**Specificity**: When the actual value is negative, how often is the prediction correct?

how often is it **CORRECTLY** predicted that the mushroom is **NOT POISONOUS**?

* Something we want to maximize
* How "specific" is the classifier in predicting positive instances?
* TN / all negative
    * all negative = TN + FP

In [19]:
specificity = TN / float(TN + FP)

print(specificity)

0.925609756097561


**False Positive Rate**: When the actual value is negative, how often is the prediction incorrect?

### how often is it **INCORRECTLY** predicted that the mushroom is **NOT POISONOUS**?

* FP / all negative

In [20]:
false_positive_rate = FP / float(TN + FP)

print(false_positive_rate)
print(1 - specificity)

0.07439024390243902
0.07439024390243898


**Precision**: When a positive value is predicted, how often is the prediction correct?

### how often is it **CORRECTLY** predicted that the mushroom is predicted to be **POISONOUS**?

How "precise" is the classifier when predicting positive instances?

In [21]:
precision = TP / float(TP + FP)

print(precision)
print(metrics.precision_score(y_test, y_pred_class))

0.922979797979798
0.922979797979798
